File: android/vbt-app/app/src/main/java/com/vbt/app/ui/screen/home/HomeViewModel.kt
package com.vbt.app.ui.screen.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vbt.app.data.remote.ApiService
import com.vbt.app.data.remote.CalendarEntryDto
import com.vbt.app.data.remote.DashboardStatsDto
import com.vbt.app.data.remote.RecentSessionDto
import com.vbt.app.data.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import javax.inject.Inject

data class HomeUiState(
    val username: String = "",
    val role: String = "athlete",
    val isBleConnected: Boolean = false,
    val todayEntries: List<CalendarEntryDto> = emptyList(),
    val dashboardStats: DashboardStatsDto? = null,
    val recentSessions: List<RecentSessionDto> = emptyList(),
    val isLoading: Boolean = false,
    // Ustawiane, gdy którekolwiek zapytanie do serwera (kalendarz/statystyki/
    // ostatnie sesje) nie powiodło się z powodu braku sieci - dashboard i tak
    // pokazuje to, co udało się wczytać wcześniej/lokalnie, ale użytkownik musi
    // wiedzieć, że dane mogą być nieaktualne zamiast po cichu widzieć puste karty.
    val offlineNotice: String? = null
)

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val apiService: ApiService
) : ViewModel() {

    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()

    val isCoach: Boolean
        get() = _uiState.value.role == "coach"

    init {
        loadData()
    }

    fun loadData() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                combine(
                    authRepository.getCurrentUsername(),
                    authRepository.getCurrentRole()
                ) { username, role ->
                    Pair(username ?: "", role ?: "athlete")
                }.collect { (username, role) ->
                    _uiState.value = _uiState.value.copy(
                        username = username,
                        role = role
                    )
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }

            var anyNetworkFailure = false

            val today = LocalDate.now().format(DateTimeFormatter.ISO_DATE)
            try {
                val calendarResponse = apiService.getCalendarEntries(dateStart = today, dateEnd = today)
                if (calendarResponse.isSuccessful) {
                    _uiState.value = _uiState.value.copy(todayEntries = calendarResponse.body() ?: emptyList())
                } else {
                    anyNetworkFailure = true
                }
            } catch (e: Exception) {
                anyNetworkFailure = true
                e.printStackTrace()
            }

            try {
                val statsResponse = apiService.getDashboardStats()
                if (statsResponse.isSuccessful) {
                    _uiState.value = _uiState.value.copy(dashboardStats = statsResponse.body())
                } else {
                    anyNetworkFailure = true
                }
            } catch (e: Exception) {
                anyNetworkFailure = true
                e.printStackTrace()
            }

            try {
                val recentResponse = apiService.getRecentSessions(limit = 5)
                if (recentResponse.isSuccessful) {
                    _uiState.value = _uiState.value.copy(recentSessions = recentResponse.body() ?: emptyList())
                } else {
                    anyNetworkFailure = true
                }
            } catch (e: Exception) {
                anyNetworkFailure = true
                e.printStackTrace()
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                offlineNotice = if (anyNetworkFailure) "Brak połączenia z serwerem - dane mogą być nieaktualne" else null
            )
        }
    }

    fun updateBleStatus(connected: Boolean) {
        _uiState.value = _uiState.value.copy(isBleConnected = connected)
    }

    fun logout(onLogout: () -> Unit) {
        viewModelScope.launch {
            try {
                authRepository.logout()
                onLogout()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}

File: ios/VBT/VBT/Features/Home/HomeViewModel.swift
import Foundation

/// Port 1:1 z Android `HomeViewModel.kt`.
@Observable
final class HomeViewModel {
    private(set) var username = ""
    private(set) var role = "athlete"
    var isBleConnected = false
    private(set) var todayEntries: [CalendarEntryDto] = []
    private(set) var dashboardStats: DashboardStatsDto?
    private(set) var recentSessions: [RecentSessionDto] = []
    private(set) var isLoading = false
    /// Ustawiane, gdy którekolwiek zapytanie do serwera nie powiodło się z powodu braku
    /// sieci — dashboard i tak pokazuje to, co udało się wczytać wcześniej, ale
    /// użytkownik musi wiedzieć, że dane mogą być nieaktualne.
    private(set) var offlineNotice: String?

    var isCoach: Bool { role == "coach" }

    private let authRepository: AuthRepository
    private let apiClient: APIClient

    init(authRepository: AuthRepository, apiClient: APIClient) {
        self.authRepository = authRepository
        self.apiClient = apiClient
        if let user = authRepository.currentUser {
            username = user.username
            role = user.role
        }
    }

    @MainActor
    func loadData() async {
        isLoading = true
        var anyNetworkFailure = false

        let today = Self.dateOnlyFormatter.string(from: Date())

        do {
            todayEntries = try await apiClient.send(.getCalendarEntries(dateStart: today, dateEnd: today))
        } catch {
            anyNetworkFailure = true
        }

        do {
            dashboardStats = try await apiClient.send(.dashboardStats())
        } catch {
            anyNetworkFailure = true
        }

        do {
            recentSessions = try await apiClient.send(.recentSessions(limit: 5))
        } catch {
            anyNetworkFailure = true
        }

        isLoading = false
        offlineNotice = anyNetworkFailure ? "Brak połączenia z serwerem - dane mogą być nieaktualne" : nil
    }

    func logout() {
        authRepository.logout()
    }
}

private extension ISO8601DateFormatter {
    static let dateOnly: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.calendar = Calendar(identifier: .iso8601)
        return formatter
    }()
}
